blockscout-ens/bens-logic/src/protocols/protocoler.rs

```diff
@@ -283,16 +283,94 @@ impl Protocoler {
         Ok(protocols)
     }
 
+    pub fn fetch_domain_options(
+        &self,
+        name_with_tld: &str,
+        network_id: i64,
+        maybe_filter: Option<NonEmpty<String>>,
+    ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
+        let tld = self.extract_tld(name_with_tld)?;
+
+        let protocols = self.protocols_of_network_for_tld(network_id, tld, maybe_filter)?;
+        let domain_names: Vec<_> = protocols
+            .into_iter()
+            .filter_map(|p| DomainNameOnProtocol::from_str(name_with_tld, p).ok())
+            .collect();
+
+        if domain_names.is_empty() {
+            return Err(ProtocolError::InvalidName {
+                name: name_with_tld.to_string(),
+                reason: "No matching protocols for given TLD".to_string(),
+            });
+        }
+
+        Ok(domain_names)
+    }
+
+    pub fn lookup_names_options_in_network(
+        &self,
+        name: &str,
+        network_id: i64,
+        maybe_filter: Option<NonEmpty<String>>,
+    ) -> Result<Vec<DomainNameOnProtocol>, ProtocolError> {
+        let protocols = self
+            .networks
+            .get(&network_id)
+            .ok_or_else(|| ProtocolError::NetworkNotFound(network_id))?
+            .use_protocols
+            .iter()
+            .filter_map(|proto_name| self.protocols.get(proto_name))
+            .collect::<Vec<_>>();
+
+        for proto in &protocols {
```

Comment:
it looks like you dont need protocol here. you pass raw string to fetch_domain_options, therefore you can just create DomainName without protocol and pass raw string. I see it like

```
fn lookup_names_options_in_network() {
    // comment that describes why you use None here
    let domain = DomainName::new(name, None)?;
    if domain.level_gt_tld() {
         // handle case of `vitalik.eth`
    }  else if domain.name.len() > 2  {
        // handle case of `eth`
} else {
   // handle case of `aa`
 }
}
```